module/zfs/ddt.c

```diff
@@ -816,14 +816,14 @@ ddt_entry_compare(const void *x1, const void *x2)
 {
 	const ddt_entry_t *dde1 = x1;
 	const ddt_entry_t *dde2 = x2;
-	const uint64_t *u1 = (const uint64_t *)&dde1->dde_key;
-	const uint64_t *u2 = (const uint64_t *)&dde2->dde_key;
+	const ddt_key_cmp_t *u1 = (ddt_key_cmp_t *)&dde1->dde_key;
+	const ddt_key_cmp_t *u2 = (ddt_key_cmp_t *)&dde2->dde_key;
 	int i;
 
 	for (i = 0; i < DDT_KEY_WORDS; i++) {
-		if (u1[i] < u2[i])
+		if (u1->dkc_key[i] < u2->dkc_key[i])
 			return (-1);
-		if (u1[i] > u2[i])
+		if (u1->dkc_key[i] > u2->dkc_key[i])
 			return (1);
 	}
```

Comment:
nit:
The common case is return != 0 (we see more ddt entries than ddt matches). Because dck_key[0] is in the checksum, differences benefit from the avalanche effect. So a slight unrolling could save a cycle or three. In pseudo-code:

if (u1->dkc_key[0] < u2->dkc_key[0])
  return (-1)
if (u1->dck_key[0] > u2->dkc_key[0])
  return (1)
for (i=1; i < DDT_KEY_WORDS && u1->dck_key[i] == u2->dkc_key[i]; i++)
  ;
if (i >= DDT_KEY_WORDS)
  return (0)
if (u1->dkc_key[i] < u2->dkc_key[i])
  return (-1)
return (1)